File: frontend/src/api/devices/lockScreen.ts
/**
 * Lock-screen access — the opt-in.
 *
 * OFF UNLESS SOMEBODY TURNS IT ON. Installing Puca registers no service,
 * creates nothing under Program Files, and raises no elevation prompt. This
 * module is the only route to any of that, and it exists behind a switch the
 * owner has to find and flip. A LocalSystem service arriving with a routine app
 * update is precisely what people are right to be suspicious of, and shipping
 * one that way would be indefensible however useful it is.
 *
 * WHAT TURNING IT ON ACTUALLY DOES, so the UI can say it plainly:
 *   - copies two binaries into %ProgramFiles%\Puca\service with a
 *     protected, administrator-only ACL;
 *   - registers a Windows service, `SovereignRemote`, running as LocalSystem
 *     and starting at boot;
 *   - that service runs an agent ONLY while the machine is locked or nobody is
 *     signed in, and stops it again on unlock.
 *
 * Turning it off removes all three. "Off" has to be a state, not a claim.
 */
import { isTauri } from '../platform';
import { apiClient } from '../client';
import { API_BASE_URL } from '../config';
import { getToken, decodeJwtPayload } from '../auth';
import { getActiveIdentity, deriveAccountSigningKey } from '../e2ee';
import { buildAuthRecord, signAuthRecord } from './identity';
import { buildUaRecord } from './unattended';

async function invokeTauri<T>(cmd: string, args?: Record<string, unknown>): Promise<T> {
    const { invoke } = await import('@tauri-apps/api/core');
    return invoke<T>(cmd, args);
}

export interface LockScreenState {
    /** The service is registered on this machine. */
    installed: boolean;
    /** Registered AND currently running. */
    running: boolean;
    /** This build ships the components at all. */
    available: boolean;
    /** Why it cannot be offered, when it cannot. */
    problem?: string | null;
}

/** Only the desktop app can install a Windows service. */
export function lockScreenSupported(): boolean {
    return isTauri();
}

/**
 * Read the REAL state from Windows, never a stored boolean.
 *
 * The service can be removed with `sc delete`, by another admin, or by an
 * uninstall this app never saw. A remembered flag would then show a toggle that
 * is on while nothing is installed — which is the same class of bug as a
 * settings screen that reports what it wrote rather than what is true.
 */
export async function lockScreenState(): Promise<LockScreenState> {
    if (!isTauri()) {
        return { installed: false, running: false, available: false, problem: null };
    }
    try {
        return await invokeTauri<LockScreenState>('service_state');
    } catch (e) {
        return {
            installed: false,
            running: false,
            available: false,
            problem: e instanceof Error ? e.message : 'Could not read the service state.',
        };
    }
}

/**
 * Turn it on. Raises exactly one Windows elevation prompt.
 *
 * Returns an error message rather than throwing, because every failure here is
 * something the user should read: declining the prompt, a missing component, a
 * refused install path. A silent revert would leave them staring at a switch
 * that snapped back for no stated reason.
 */
export async function enableLockScreenAccess(): Promise<string | null> {
    if (!isTauri()) return 'This can only be turned on from the desktop app.';
    try {
        await invokeTauri<void>('service_enable');
        return null;
    } catch (e) {
        return e instanceof Error ? e.message : String(e);
    }
}

/** Turn it off: stop the service, delete it, remove its files. */
export async function disableLockScreenAccess(): Promise<string | null> {
    if (!isTauri()) return 'This can only be turned off from the desktop app.';
    try {
        await invokeTauri<void>('service_disable');
        return null;
    } catch (e) {
        return e instanceof Error ? e.message : String(e);
    }
}

/**
 * The two switches, read from the machine rather than remembered.
 *
 * ENROLLED means this computer can be reached at its sign-in screen at all.
 * ARMED means a session there may be authorised by proving a passphrase. Both
 * are required before anything happens, and they are separate so that disarming
 * can shut the door without tearing down the connection you would need to
 * re-arm it.
 */
export interface UnattendedAccessState {
    serviceInstalled: boolean;
    enrolled: boolean;
    armed: boolean;
    /**
     * The device row this machine is enrolled as at its sign-in screen, or null.
     *
     * ONE PC IS TWO DEVICE ROWS — the app's and this service's, each with its
     * own keypair because neither process may hold the other's. Nothing linked
     * them, so the list showed the same machine twice and, worse, only the
     * app's row ever had a MAC recorded: the row you can actually reach while
     * the screen is locked could never be woken. `lanInfo.ts` uses this id to
     * publish the same sealed LAN details to both rows, which fixes the wake
     * and makes the two rows recognisable as one machine (same MAC).
     */
    deviceId: string | null;
    /**
     * Fingerprint of the INSTALLED service+agent pair, reported by the running
     * service itself. Null from an old service that predates the field — which
     * is itself the strongest possible "this service needs updating".
     */
    binsHash: string | null;
    error?: string | null;
}

export async function unattendedAccessState(): Promise<UnattendedAccessState> {
    if (!isTauri()) {
        return {
            serviceInstalled: false, enrolled: false, armed: false,
            deviceId: null, binsHash: null,
        };
    }
    try {
        const s = await invokeTauri<{
            service_installed: boolean;
            enrolled: boolean;
            armed: boolean;
            device_id?: string | null;
            bins_hash?: string | null;
            error?: string | null;
        }>('lock_screen_state');
        return {
            serviceInstalled: s.service_installed,
            enrolled: s.enrolled,
            armed: s.armed,
            deviceId: s.device_id ?? null,
            binsHash: s.bins_hash ?? null,
            error: s.error ?? null,
        };
    } catch (e) {
        return {
            serviceInstalled: false,
            enrolled: false,
            armed: false,
            deviceId: null,
            binsHash: null,
            error: e instanceof Error ? e.message : String(e),
        };
    }
}

/**
 * Does the running service predate the app that is asking?
 *
 * A PURE decision so it can be tested: `reported` is what the running service
 * says it is (null from a service too old to say — which IS out of date), and
 * `bundled` is what this app ships (null in a dev build with no sidecar, where
 * no update can be offered anyway).
 *
 * WHY THIS MATTERS ENOUGH TO EXIST: the app auto-updates and the service does
 * not — nothing but enrolment day ever touched it. The skew is invisible: the
 * pipe still answers, just without whatever fields the newer app relies on.
 * That is exactly how 0.8.82's one-card-per-PC merge silently never engaged —
 * the service never sent `device_id`, so the sign-in row never got its MAC.
 */
export function serviceNeedsUpdate(
    reported: string | null,
    bundled: string | null,
): boolean {
    if (!bundled) return false;
    return reported !== bundled;
}

/**
 * Should the app-root "update the service" banner show? The card's decision
 * plus the one thing the card's placement already implied: there has to BE a
 * service. `installed` from the SCM (`service_state`), `reported` from the
 * running service over the control pipe, `bundled` from this build's
 * sidecars. Pure; the banner (`ServiceUpdateBanner.tsx`) is the only caller.
 */
export function serviceUpdateBannerDue(
    installed: boolean,
    reported: string | null,
    bundled: string | null,
): boolean {
    return installed && serviceNeedsUpdate(reported, bundled);
}

export interface BundledFingerprint {
    hash: string | null;
    /**
     * Set only when the pair SHOULD have been readable and was not — never
     * for the ordinary "this build has no sidecars" case, which is `hash:
     * null, error: null`. The distinction matters: collapsing both into a
     * bare `null` is exactly how the previous version of this went missing
     * with nothing anywhere saying why — on a real install, `bundled` being
     * silently null makes `serviceNeedsUpdate` return false and the "Update
     * the service" card never appears, indistinguishable from "already
     * current". This field is what turns that into something reportable.
     */
    error: string | null;
}

/** The fingerprint of the service+agent pair bundled with THIS app build. */
export async function bundledServiceFingerprint(): Promise<BundledFingerprint> {
    if (!isTauri()) return { hash: null, error: null };
    try {
        const hash = await invokeTauri<string | null>('service_bundled_fingerprint');
        return { hash, error: null };
    } catch (e) {
        return { hash: null, error: e instanceof Error ? e.message : String(e) };
    }
}

/**
 * Replace the installed service binaries with this build's. One elevation
 * prompt; the registration, enrolment and passphrase all survive.
 */
export async function updateLockScreenService(): Promise<string | null> {
    if (!isTauri()) return 'This can only be done from the desktop app.';
    try {
        await invokeTauri<void>('service_update');
        return null;
    } catch (e) {
        return e instanceof Error ? e.message : String(e);
    }
}

/**
 * Enrol this machine so its sign-in screen can be reached.
 *
 * THREE STEPS, AND THE ORDER IS THE POINT. The service generates its own device
 * keypair and hands back only the public halves; this signs a record describing
 * them with the ACCOUNT key and enrols it with the server; only then is the
 * service told where to connect and which account key to verify others against.
 *
 * The account signing key never reaches the service. It is derived from the
 * account seed, and a LocalSystem process holding an internet socket must not be
 * able to enrol devices into the account — it gets the PUBLIC half, which is
 * enough to verify a peer and useless for enrolling anything.
 *
 * Returns an error message rather than throwing, matching the other toggles in
 * this file, because every failure here is something the user needs to read.
 */
export async function enrolLockScreenAccess(name?: string): Promise<string | null> {
    if (!isTauri()) return 'This can only be turned on from the desktop app.';

    const identity = getActiveIdentity();
    if (!identity) {
        return 'Unlock Puca on this computer first — enrolling has to sign with your account key.';
    }
    const token = getToken();
    if (!token) return 'Sign in first: this computer needs a copy of the session to connect.';

    // CAPTURE THE PREDECESSOR'S ROW ID BEFORE IT IS UNKNOWABLE. `begin_enrol`
    // overwrites the stored keys, and the device id is derived from the keys —
    // so the old row's id exists nowhere after that call. Every re-enrolment
    // used to orphan the previous row this way: permanently offline (its
    // private key is destroyed), never grouped, never cleaned up. Three of
    // them accumulated on one machine before anyone understood why.
    let priorDeviceId: string | null = null;
    try {
        const prior = await unattendedAccessState();
        priorDeviceId = prior.enrolled ? prior.deviceId : null;
    } catch {
        // Unknowable is survivable: worst case is the old behaviour, one
        // orphan row the user can revoke by hand.
    }

    // The account id comes from the TOKEN, not from the identity: `Identity`
    // carries keys, not an account number, and the server derives the same id
    // from the same token when it validates the enrolment.
    let userId: number;
    try {
        const sub = decodeJwtPayload(token)?.sub;
        if (typeof sub !== 'number') throw new Error('no sub');
        userId = sub;
    } catch {
        return 'Could not tell which account this is. Sign in again and retry.';
    }

    try {
        // 1. The service makes its own keys. The private halves never leave it.
        const keys = await invokeTauri<{
            device_id: string;
            device_pub: string;
            sign_pub: string;
        }>('lock_screen_begin_enrol');

        // 2. Sign a record describing them, and enrol it.
        const deviceName = name?.trim() || 'This PC (sign-in screen)';
        const { canonical, deviceId } = buildAuthRecord({
            devicePub: keys.device_pub,
            signPub: keys.sign_pub,
            name: deviceName,
            platform: 'windows',
            userId,
        });

        // The server derives the id independently and ignores what we claim, so
        // a mismatch means the service handed back keys that do not hash to the
        // id it reported. Catch it now rather than shipping a device row that
        // can never attest.
        if (deviceId !== keys.device_id) {
            return 'This computer reported keys that do not match its own device id. Try again.';
        }

        await apiClient.post('/devices', {
            device_pub: keys.device_pub,
            sign_pub: keys.sign_pub,
            name: deviceName,
            platform: 'windows',
            auth_record: canonical,
            auth_sig: signAuthRecord(identity, canonical),
        });

        // 3. Tell the service where to connect, and whose signature to trust.
        await invokeTauri<void>('lock_screen_finish_enrol', {
            apiBase: API_BASE_URL,
            userId,
            token,
            accountSignPub: deriveAccountSigningKey(identity).publicKeyEncoded,
        });

        // 4. Retire the predecessor row — ONLY now, with the new enrolment
        // fully landed. Revoking first would leave the machine unreachable if
        // any later step failed. Best-effort: its key is already destroyed,
        // so the worst a failure here costs is one stale card.
        if (priorDeviceId && priorDeviceId !== keys.device_id) {
            try {
                const { revokeDevice } = await import('./index');
                await revokeDevice(priorDeviceId);
            } catch {
                // The old row stays visible until revoked by hand. Not worth
                // failing an enrolment that succeeded.
            }
        }

        // 5. Give the fresh row its LAN details NOW. The steady-state publish
        // fires on the next socket attestation — days away on a desktop that
        // stays up — and until then this row would be un-wakeable and
        // un-grouped, which is precisely the field failure this fixes.
        try {
            const { publishNow } = await import('./lanInfo');
            await publishNow();
        } catch {
            // The attestation-day publish remains as the retry path.
        }
        return null;
    } catch (e) {
        return e instanceof Error ? e.message : String(e);
    }
}

/** Forget everything: keys, token, connection details and the passphrase —
 *  and retire the server row, which "forget" used to leave behind as a
 *  permanently-offline card (the keys that could attest for it are gone). */
export async function unenrolLockScreenAccess(): Promise<string | null> {
    if (!isTauri()) return 'This can only be turned off from the desktop app.';

    // Captured BEFORE the service forgets: afterwards nothing knows the id.
    let priorDeviceId: string | null = null;
    try {
        const prior = await unattendedAccessState();
        priorDeviceId = prior.enrolled ? prior.deviceId : null;
    } catch {
        // Unknowable is survivable — one stale card, removable by hand.
    }

    try {
        await invokeTauri<void>('lock_screen_unenrol');
    } catch (e) {
        return e instanceof Error ? e.message : String(e);
    }

    if (priorDeviceId) {
        try {
            const { revokeDevice } = await import('./index');
            await revokeDevice(priorDeviceId);
        } catch {
            // The machine-side forget succeeded, which is the half that
            // matters for security; the leftover card is cosmetic.
        }
    }
    return null;
}

/**
 * Set the passphrase that authorises a session at the sign-in screen.
 *
 * Derived HERE; only the salt and the public verifying key are stored on the
 * machine. The passphrase itself exists inside this call and nowhere else — not
 * returned, not logged, not kept.
 *
 * THIS IS THE ONLY THING PROTECTING A MACHINE THAT IS SITTING AT ITS LOGIN
 * SCREEN, and there is no server-side rate limit behind it and no recovery, so
 * the minimum length is enforced here rather than only in the UI.
 */
export async function armLockScreenAccess(passphrase: string): Promise<string | null> {
    if (!isTauri()) return 'This can only be set from the desktop app.';
    if (passphrase.length < 8) {
        return 'Use at least 8 characters — this is the only thing protecting the sign-in screen.';
    }
    try {
        const record = buildUaRecord(passphrase);
        await invokeTauri<void>('lock_screen_arm', { record: JSON.stringify(record) });
        return null;
    } catch (e) {
        return e instanceof Error ? e.message : String(e);
    }
}

/** Shut the door. Removing the record IS the revocation — the agent re-reads
 *  it on every connection, so this takes effect on the next session and not at
 *  the next reboot. */
export async function disarmLockScreenAccess(): Promise<string | null> {
    if (!isTauri()) return 'This can only be changed from the desktop app.';
    try {
        await invokeTauri<void>('lock_screen_disarm');
        return null;
    } catch (e) {
        return e instanceof Error ? e.message : String(e);
    }
}

File: frontend/src-tauri/src/lock_screen.rs
//! Turning sign-in-screen access on, and off.
//!
//! WHAT THIS IS FOR. Every other remote-control path in this app needs the app
//! running, which means somebody has already signed in. This one is for the
//! machine that is sitting at its Windows sign-in screen — after a reboot, or
//! after being woken from off — where there is no app and no signed-in user.
//! Only the SYSTEM service can reach that, and only if the owner deliberately
//! turned it on.
//!
//! OFF BY DEFAULT, AND ABSENT RATHER THAN DISABLED. A machine that never enrols
//! has no config, no keys and no token, so the service's link thread finds
//! nothing and never opens a socket. `forget` puts it back exactly as it was.
//!
//! TWO SEPARATE SWITCHES, deliberately:
//!   * ENROLLED — this machine can be reached at its sign-in screen at all;
//!   * ARMED — a session there may be authorised by proving a passphrase.
//! Both are required for anything to happen. They are separate so that
//! disarming can shut the door without also tearing down the connection you
//! would need in order to re-arm it.
//!
//! WHY ENROLMENT IS TWO CALLS. Publishing a device into the account needs an
//! auth record signed by the ACCOUNT signing key, which is derived from the
//! account seed and must never reach a LocalSystem process holding an internet
//! socket. So the service generates the keypair and hands back only public
//! halves, the web layer signs a record describing them and enrols it, and the
//! service is then told the account's PUBLIC key — enough to verify other
//! devices, useless for enrolling any.

#![cfg(windows)]

use serde::Serialize;
use puca_service::control::{ControlRequest, CONTROL_PIPE};
use std::io::{BufRead, BufReader, Write};

/// Send one request to the service and read its reply.
///
/// Every call opens and closes the pipe. The service's control pipe is
/// request/response with no session state, and holding one open would mean a
/// UI toggle pinning a handle to a service that may restart under it.
fn ask(req: &ControlRequest) -> Result<serde_json::Value, String> {
    use std::os::windows::fs::OpenOptionsExt;
    // SECURITY_IDENTIFICATION, for the reason service_link::ask gives.
    const SECURITY_SQOS_PRESENT: u32 = 0x0010_0000;
    const SECURITY_IDENTIFICATION: u32 = 0x0001_0000;
    let file = std::fs::OpenOptions::new()
        .read(true)
        .write(true)
        .custom_flags(SECURITY_SQOS_PRESENT | SECURITY_IDENTIFICATION)
        .open(CONTROL_PIPE)
        .map_err(|_| "the Puca service is not installed on this computer".to_string())?;

    let mut writer = file
        .try_clone()
        .map_err(|e| format!("could not talk to the service: {e}"))?;
    let line = serde_json::to_string(req).map_err(|e| format!("could not encode: {e}"))?;
    writeln!(writer, "{line}").map_err(|e| format!("could not talk to the service: {e}"))?;
    writer.flush().ok();

    let mut reply = String::new();
    BufReader::new(file)
        .read_line(&mut reply)
        .map_err(|e| format!("the service did not answer: {e}"))?;

    let v: serde_json::Value = serde_json::from_str(reply.trim())
        .map_err(|_| "the service sent something unreadable".to_string())?;

    // A refusal carries the reason the service already worded for a person —
    // "only the user signed in at this computer's own screen can do this" — and
    // replacing it with a generic failure here would throw away the only part
    // that tells the user what to do.
    if v.get("t").and_then(|x| x.as_str()) == Some("refused") {
        return Err(v
            .get("reason")
            .and_then(|x| x.as_str())
            .unwrap_or("the service refused, without saying why")
            .to_string());
    }
    Ok(v)
}

#[derive(Serialize, Default)]
pub struct LockScreenState {
    /// Is the service present at all? Everything else is meaningless without it.
    pub service_installed: bool,
    pub enrolled: bool,
    pub armed: bool,
    /// The device row this machine is enrolled as at its sign-in screen, or
    /// None when it is not enrolled.
    ///
    /// THE APP USES THIS TO STOP SHOWING ONE PC AS TWO DEVICES, and to PATCH
    /// this row's LAN details so it can actually be woken — see
    /// `puca_service::link::enrolled_device_id`, which carries the full
    /// reasoning, and `frontend/src/api/devices/lanInfo.ts`, which consumes it.
    pub device_id: Option<String>,
    /// Fingerprint of the INSTALLED service+agent pair, reported by the
    /// running service itself. Compared against `service_bundled_fingerprint`
    /// to decide whether the running service predates this app — the app
    /// auto-updates, the service does not, and the skew is invisible until a
    /// control-pipe field the app relies on silently never arrives. Absent
    /// from an older service, which is itself the strongest "needs update".
    pub bins_hash: Option<String>,
    /// What the SERVER last said about the sign-in-screen link, from the
    /// service's link-health record (`puca_service::link::LinkHealth`), for
    /// the enrolled identity only. `enrolled` above is local files and reads
    /// true while the server refuses the machine; these are what can tell.
    /// All None from an older service, which the UI reads as "nothing to say".
    /// Unix seconds.
    pub link_attested_at: Option<i64>,
    pub link_refused_first: Option<i64>,
    pub link_refused_last: Option<i64>,
    pub link_refused_count: u32,
    pub link_refused_status: Option<u16>,
    /// Why the state could not be read, if it could not. Shown rather than
    /// swallowed: a toggle that silently renders "off" when it could not ask is
    /// how someone concludes a feature is broken.
    pub error: Option<String>,
}

/// Read the real state, never a stored flag.
///
/// The answer lives in files the service owns and that an administrator can
/// remove by hand. A cached boolean here would eventually claim a machine is
/// armed when its record has been deleted.
#[tauri::command]
pub async fn lock_screen_state() -> LockScreenState {
    tauri::async_runtime::spawn_blocking(|| match ask(&ControlRequest::UnattendedState) {
        Ok(v) => read_unattended(&v),
        Err(e) => LockScreenState { error: Some(e), ..LockScreenState::default() },
    })
    .await
    .unwrap_or_default()
}

/// Pull the three facts out of the service's reply.
///
/// A NAMED FUNCTION so a test can feed it a REAL, serialised
/// `ControlResponse::UnattendedState` built from the service's own type. These
/// two crates compile separately and are joined only by these key strings, and
/// this repo has repeatedly shipped cross-process names that failed silently —
/// a wrong key here would simply report an unenrolled machine for ever, which
/// looks exactly like a machine nobody turned this on for.
fn read_unattended(v: &serde_json::Value) -> LockScreenState {
    LockScreenState {
        service_installed: true,
        enrolled: v.get("enrolled").and_then(|x| x.as_bool()).unwrap_or(false),
        armed: v.get("armed").and_then(|x| x.as_bool()).unwrap_or(false),
        // Absent from an older service, and null when unenrolled — both mean
        // "no companion row to pair with", which is the safe answer.
        device_id: v
            .get("device_id")
            .and_then(|x| x.as_str())
            .filter(|s| !s.is_empty())
            .map(str::to_string),
        // Absent (old service) reads as None, which the UI treats as "needs
        // update" — an old service is precisely a service needing an update.
        bins_hash: v
            .get("bins_hash")
            .and_then(|x| x.as_str())
            .filter(|s| !s.is_empty())
            .map(str::to_string),
        link_attested_at: link_i64(v, "attested_at"),
        link_refused_first: link_i64(v, "refused_first"),
        link_refused_last: link_i64(v, "refused_last"),
        link_refused_count: v
            .get("link")
            .and_then(|l| l.get("refused_count"))
            .and_then(|x| x.as_u64())
            .map(|n| n.min(u32::MAX as u64) as u32)
            .unwrap_or(0),
        // Only a client-error status is a refusal; anything else in this slot
        // is a service this app does not understand, and says nothing.
        link_refused_status: v
            .get("link")
            .and_then(|l| l.get("refused_status"))
            .and_then(|x| x.as_u64())
            .filter(|s| (400..=499).contains(s))
            .map(|s| s as u16),
        error: None,
    }
}

/// One timestamp from the service's `link` object. Absent (an older service,
/// or nothing recorded) is None.
fn link_i64(v: &serde_json::Value, key: &str) -> Option<i64> {
    v.get("link").and_then(|l| l.get(key)).and_then(|x| x.as_i64())
}

#[derive(Serialize)]
pub struct BegunEnrolment {
    pub device_id: String,
    pub device_pub: String,
    pub sign_pub: String,
}

/// Step one: ask the service to generate this machine's device identity.
///
/// Returns PUBLIC halves only. The caller signs an auth record over them with
/// the account key and enrols it, then calls `lock_screen_finish_enrol`.
#[tauri::command]
pub async fn lock_screen_begin_enrol() -> Result<BegunEnrolment, String> {
    tauri::async_runtime::spawn_blocking(|| {
        let v = ask(&ControlRequest::EnrolBegin)?;
        let s = |k: &str| {
            v.get(k)
                .and_then(|x| x.as_str())
                .map(str::to_string)
                .ok_or_else(|| format!("the service did not return {k}"))
        };
        Ok(BegunEnrolment {
            device_id: s("device_id")?,
            device_pub: s("device_pub")?,
            sign_pub: s("sign_pub")?,
        })
    })
    .await
    .map_err(|e| format!("the request did not finish: {e}"))?
}

/// Step two: hand the service the connection details.
#[tauri::command]
pub async fn lock_screen_finish_enrol(
    api_base: String,
    user_id: i64,
    token: String,
    account_sign_pub: String,
) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        ask(&ControlRequest::EnrolFinish { api_base, user_id, token, account_sign_pub })
            .map(|_| ())
    })
    .await
    .map_err(|e| format!("the request did not finish: {e}"))?
}

/// Arm this machine with a record derived from a passphrase.
///
/// The record is the salt and the PUBLIC verifying key — never the passphrase
/// and never the private key, which exist only on the controlling device.
#[tauri::command]
pub async fn lock_screen_arm(record: String) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        ask(&ControlRequest::Arm { record }).map(|_| ())
    })
    .await
    .map_err(|e| format!("the request did not finish: {e}"))?
}

#[tauri::command]
pub async fn lock_screen_disarm() -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(|| ask(&ControlRequest::Disarm).map(|_| ()))
        .await
        .map_err(|e| format!("the request did not finish: {e}"))?
}

/// Forget everything: keys, token, config and the arming record.
#[tauri::command]
pub async fn lock_screen_unenrol() -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(|| ask(&ControlRequest::Unenrol).map(|_| ()))
        .await
        .map_err(|e| format!("the request did not finish: {e}"))?
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn requests_serialise_the_way_the_service_parses_them() {
        // BUILT FROM THE SERVICE'S OWN TYPE. The last bug in this file was a
        // handwritten frame that no test ever ran — see service_link.rs's
        // parse_handle. Round-tripping through the real enum is what makes a
        // misspelling impossible rather than merely unlikely.
        let reqs = [
            ControlRequest::UnattendedState,
            ControlRequest::EnrolBegin,
            ControlRequest::Unenrol,
            ControlRequest::Disarm,
            ControlRequest::Arm { record: "{}".into() },
            ControlRequest::EnrolFinish {
                api_base: "https://chat.example.com".into(),
                user_id: 1,
                token: "t".into(),
                account_sign_pub: "ed25519:AAAA".into(),
            },
        ];
        for r in &reqs {
            let wire = serde_json::to_string(r).expect("serialise");
            // Tagged "t", snake_case — the shape control_pipe.rs deserialises.
            let v: serde_json::Value = serde_json::from_str(&wire).expect("parse");
            let tag = v.get("t").and_then(|x| x.as_str()).expect("a tag");
            assert_eq!(tag, tag.to_lowercase(), "tags are snake_case: {tag}");
            // And it must deserialise back into the very same request.
            let back: ControlRequest = serde_json::from_str(&wire).expect("round trip");
            assert_eq!(&back, r);
        }
    }

    #[test]
    fn a_refusal_keeps_the_services_own_wording() {
        // The service words refusals for a person. Replacing them with a
        // generic failure throws away the only part that says what to do.
        let refused = puca_service::control::ControlResponse::Refused {
            reason: "only the user signed in at this computer's own screen can do this".into(),
        };
        let wire = serde_json::to_string(&refused).unwrap();
        let v: serde_json::Value = serde_json::from_str(&wire).unwrap();
        assert_eq!(v.get("t").and_then(|x| x.as_str()), Some("refused"));
        assert_eq!(
            v.get("reason").and_then(|x| x.as_str()),
            Some("only the user signed in at this computer's own screen can do this")
        );
    }

    #[test]
    fn no_service_is_reported_as_absent_rather_than_as_a_failure() {
        // The default on every machine that never turned this on. It must not
        // read as an error the user has to act on — and the state struct's
        // default must be the safe one, since spawn_blocking failure falls back
        // to it.
        let d = LockScreenState::default();
        assert!(!d.service_installed && !d.enrolled && !d.armed);
        assert!(d.device_id.is_none(), "no service means no companion row");
    }

    #[test]
    fn the_enrolled_device_id_survives_the_trip_from_the_service() {
        // BOTH ENDS, NOT A GUESSED STRING. The expectation is built from the
        // service's OWN type and serialised exactly as the pipe would carry it,
        // then handed to the real parser. Rename the field on either side and
        // this fails — whereas the product would merely go on reporting "no
        // companion row", i.e. one PC listed as two devices with the sign-in
        // half permanently un-wakeable, and nothing would say why.
        let reply = puca_service::control::ControlResponse::UnattendedState {
            armed: true,
            enrolled: true,
            device_id: Some("iE7UN9h775LJJ1rIbNblf".into()),
            bins_hash: Some("abc123".into()),
            link: Some(puca_service::control::LinkHealthView {
                attested_at: Some(1),
                refused_first: Some(2),
                refused_last: Some(3),
                refused_count: 4,
                refused_status: Some(400),
            }),
        };
        let v: serde_json::Value =
            serde_json::from_str(&serde_json::to_string(&reply).unwrap()).unwrap();

        let state = read_unattended(&v);
        assert_eq!(state.device_id.as_deref(), Some("iE7UN9h775LJJ1rIbNblf"));
        assert!(state.enrolled && state.armed && state.service_installed);
        // The pair fingerprint rides the same reply — BOTH ends, same test
        // discipline as device_id: a rename on either side must go red here,
        // because the product failure it causes is a permanently missing
        // "update the service" prompt, which looks like nothing at all.
        assert_eq!(state.bins_hash.as_deref(), Some("abc123"));
        // The link-health record, same discipline: a rename on either side
        // would leave the app showing a ticked box with no warning while the
        // server refuses the machine — the exact failure it was added for.
        assert_eq!(state.link_attested_at, Some(1));
        assert_eq!(state.link_refused_first, Some(2));
        assert_eq!(state.link_refused_last, Some(3));
        assert_eq!(state.link_refused_count, 4);
        assert_eq!(state.link_refused_status, Some(400));
    }

    #[test]
    fn an_unenrolled_machine_reports_no_companion_row_rather_than_a_blank_one() {
        // An empty string would sail through `Option` and become a device id
        // that PATCHes nothing and groups with nothing. Both shapes the service
        // can legitimately produce must land as None.
        let absent = puca_service::control::ControlResponse::UnattendedState {
            armed: false,
            enrolled: false,
            device_id: None,
            bins_hash: None,
            link: None,
        };
        let v: serde_json::Value =
            serde_json::from_str(&serde_json::to_string(&absent).unwrap()).unwrap();
        assert!(read_unattended(&v).device_id.is_none());

        // And an older service, which does not send the field at all.
        let old: serde_json::Value =
            serde_json::from_str(r#"{"t":"unattended_state","armed":false,"enrolled":true}"#)
                .unwrap();
        assert!(read_unattended(&old).device_id.is_none());
        assert!(read_unattended(&old).enrolled, "the rest must still parse");
        // An old service reports no fingerprint — which must land as None,
        // the value the UI reads as "this service needs updating". True by
        // construction: a service too old to say what it is IS out of date.
        assert!(read_unattended(&old).bins_hash.is_none());
        // An old service has nothing to say about the link — which must read
        // as no refusal, so the app shows exactly what it always did.
        let o = read_unattended(&old);
        assert_eq!(
            (o.link_attested_at, o.link_refused_first, o.link_refused_last),
            (None, None, None)
        );
        assert_eq!((o.link_refused_count, o.link_refused_status), (0, None));
        // As must a service that sent the key with nothing in it.
        let n = read_unattended(&v);
        assert_eq!((n.link_refused_first, n.link_refused_count), (None, 0));

        // An empty string is not an id.
        let blank: serde_json::Value = serde_json::from_str(
            r#"{"t":"unattended_state","armed":false,"enrolled":true,"device_id":""}"#,
        )
        .unwrap();
        assert!(read_unattended(&blank).device_id.is_none());
    }
}
